### ISSUE TITLE
[HOLD for payment 2023-11-21] [$500] LHN - Money Request removal offline doesn't update LHN correctly

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#28941]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open a chat with user B
2. Request money
3. Go offline
4. Open the newly created money request report
5. Delete the request

## Expected Result:
In the LHN, the chat with user B should show the last visible message

## Actual Result:
 In the LHN, the chat with user B shows "<email> owes <amount>" until going back online

## Workaround:
Unknown 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.78.0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/c5654160-fcc4-489d-814d-2fce395c3eb8

https://github.com/Expensify/App/assets/93399543/97c79d2b-3128-4702-a36b-6f44f849bfd8

https://github.com/Expensify/App/assets/93399543/57a20123-9d92-4fb0-ab0d-fcbc2afd4519

https://github.com/Expensify/App/assets/93399543/3c9dbf7b-cfa0-4ca7-bf2d-9e4e375840e9

https://github.com/Expensify/App/assets/93399543/70e924a4-1853-459e-ac8d-abc6bddb5934

https://github.com/Expensify/App/assets/93399543/d9a5a476-6b6e-4c33-8e1e-6774764011ba

https://github.com/Expensify/App/assets/93399543/70a097f0-fc0f-43ce-936c-d011be0189f0

**Expensify/Expensify Issue URL:**
**Issue reported by:** @paultsimura
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1696501720522399

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01e4ad059b092f74b3</li>
        <li>Upwork Job ID: 1709994082221408256</li>
        <li>Last Price Increase: 2023-10-12</li>
<li>Automatic offers: </li>
<ul>
<li>tienifr | Contributor | 27265296</li>
<li>paultsimura | Reporter | 27265299</li>
</ul></li>
    </ul>
</details>

**Comment by paultsimura** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Once IOU report is deleted, the parent report's last message is not updated in LHN

### What is the root cause of that problem?

When deleting a Money Request, if it's the last request in the IOU report, we optimistically update the parent report:
https://github.com/Expensify/App/blob/1663440e2e235f2ac70495732b66dba539194440/src/libs/actions/IOU.js#L1567-L1580

We explicitly tell the `ReportActionsUtils.getLastVisibleMessage` to ignore the "IOU Preview" action: `{[parentReportAction.reportActionID]: null}`
However, this `null` is ignored here, the change was introduced by https://github.com/Expensify/App/pull/26447, the `null` scenario was not ignored before:
https://github.com/Expensify/App/blob/f6efb269a587aa664ee2a94366bb9aeedcc4cae2/src/libs/ReportActionsUtils.js#L381-L387
### What changes do you think we should make in order to solve the problem?
If the `actionsToMerge` contains explicit `null`, we should set the corresponding `updatedActionsToMerge` key to `null` instead of using the spread operator:

```diff
    if (actionsToMerge && Object.keys(actionsToMerge).length !== 0) {
        Object.keys(actionsToMerge).forEach(
-           (actionToMergeID) => (updatedActionsToMerge[actionToMergeID] = {...allReportActions[reportID][actionToMergeID], ...actionsToMerge[actionToMergeID]}),
+           (actionToMergeID) => (updatedActionsToMerge[actionToMergeID] = lodashIsNull(actionsToMerge[actionToMergeID]) ? null : {...allReportActions[reportID][actionToMergeID], ...actionsToMerge[actionToMergeID]}),
        );
    }
```

And also, for better code reuse, in `IOU.deleteMoneyRequest` we should consider passing `{[reportPreviewAction.reportActionID]: updatedReportPreviewAction}` (as a param to ReportActionsUtils.getLastVisibleMessage and ReportActionsUtils.getLastVisibleAction) instead of `{[reportPreviewAction.reportActionID]: null}`, as the `updatedReportPreviewAction` is already `null` in this scenario.
### What alternative solutions did you explore? (Optional)
We can alternatively leave the `ReportActionsUtils.getLastVisibleAction` logic untouched and pass some dummy "invisible" action instead of null. Like an action with `message.html: ''`, but that's not a preferable approach.

**Comment by dylanexpensify** [comment]
@burczu can we review these soon? 🙇‍♂️

**Comment by dylanexpensify** [comment]
bump @burczu

**Comment by burczu** [comment]
@dylanexpensify Ahh, sorry - I've missed this one... I'll take a look today!

**Comment by burczu** [comment]
@paultsimura Just to note, posting code diffs in proposals is not allowed.

**Comment by paultsimura** [comment]
@burczu it was discussed multiple times recently and the agreement was that the proposal must not consist **only of code diffs**, but the code diffs are welcomed as they help describing the changes:

> Just keep it within reason. We made that guideline because:
people would post a whole PR in a comment :facepalm:
they would not post anything else but the code. everyone has to work harder to understand the changes.

https://expensify.slack.com/archives/C01GTK53T8Q/p1694106505333079

**Comment by burczu** [comment]
@paultsimura Alright! Thanks for pointing this out ;)

**Comment by burczu** [comment]
@paultsimura and @tienifr I've just tested your solutions and they both works until we go back online. Please see what happens after that:

https://github.com/Expensify/App/assets/7771498/1db76cc2-6c56-46b4-aa56-d1d79a3d89be

I think this should be also handled in this issue.

**Comment by dylanexpensify** [comment]
@burczu mind replying to the above? 🙇‍♂️

**Comment by burczu** [comment]
@dylanexpensify I'm sorry, I was busy reviewing PR's - I'll get back to checking proposals soon.

**Comment by dylanexpensify** [comment]
Sounds good, thanks @burczu!

**Comment by burczu** [comment]
Ok, I think what @tienifr wrote make sense and I tend to agree that this can be out of scope of this issue, especially that there is an issue that should address that (https://github.com/Expensify/App/issues/28737).

In that case, I prefer the solution [proposed](https://github.com/Expensify/App/issues/28941#issuecomment-1750188994) by @tienifr because it looks simpler and more clean.

🎀 👀 🎀 C+ reviewed

**Comment by dylanexpensify** [comment]
@burczu bump on update please 🙇‍♂️

**Comment by burczu** [comment]
@dylanexpensify It's under review but it seems we have a blocker based on the @tienifr's comment: https://github.com/Expensify/App/pull/29939#issuecomment-1779600738

**Comment by dylanexpensify** [comment]
Got it, thanks for update!

**Comment by dylanexpensify** [comment]
@tienifr any update here on a resolution?

**Comment by burczu** [comment]
@dylanexpensify Based on this comment: https://github.com/Expensify/App/pull/29939#issuecomment-1797849874 there is no blocker anymore, so I'll get back to this PR and review it soon.

**Comment by dylanexpensify** [comment]
Payment time! @burczu to complete checklist

**Comment by burczu** [comment]
@dylanexpensify Sorry for the delay - I'll get back to BZ Checklists soon, after I finish reviewing my PR's and proposals.

**Comment by burczu** [comment]
Yes, I’ll be completing BZ checklists tomorrow.

**Comment by burczu** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

* [x]  [@burczu] The PR that introduced the bug has been identified. Link to the PR: 

**https://github.com/Expensify/App/pull/26447**

* [x]  [@burczu] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

**https://github.com/Expensify/App/pull/26447#discussion_r1411713551**

* [x]  [@burczu] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

**Didn't find any.**

* [x]  [@burczu] Determine if we should create a regression test for this bug.
* [x]  [@burczu] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

**Regression test proposal in a separate comment.**

* [ ]  [@dylanexpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by burczu** [comment]
**Regression test proposal**

1. Login to the App
2. Open chat with some user
3. Type any message
4. Request money in this chat
5. Go offline
6. Open money request report that's just created in step 4 and delete it
7. Go back to LHN and verify that the chat shows the correct (last) message
8. Do we agree 👍 or 👎

**Comment by flodnv** [comment]
Good call, let's ask that to Applause @dylanexpensify and close this issue

**Comment by paultsimura** [comment]
Hey @dylanexpensify, could you please check the payment on this issue?
I've looked through my Upwork, and seems like my Reporter contract remains unpaid. Thanks.

**Comment by dylanexpensify** [comment]
looking now!

**Comment by dylanexpensify** [comment]
paid!

**Comment by dylanexpensify** [comment]
omg sorry! Yes on it now!

**Comment by dylanexpensify** [comment]
@tienifr done!
